Ability to stop a network

WhoSoup opened the pull request:
Now that most of the reworks have been done, I wanted to add the ability to stop the networks again. 
* All peers are stopped
* All permanent goroutines are closed

The only special case was the limited listener. It's possible to close the listener and stop the routine but if a network is stopped immediately after being started, there's a chance the listener might not be initialized yet. In order to remedy that, there's a goroutine inside the listener loop that will close the listener if the network is shut down.

PaulBernier commented on controller_connections.go:
```diff
@@ -40,6 +40,9 @@ func (c *controller) manageOnline() {
 				c.net.prom.Outgoing.Set(float64(c.peers.Outgoing()))
 			}
+		case <-c.net.stopper:
+			// ordered second so that manageOnline will clear out all peers from the peer store first
```
Are you saying `c.peerStatus` will be drained before going to the stopper? Are you sure?

https://www.reddit.com/r/golang/comments/4ibs68/why_is_the_select_statement_nondeterministic/
https://stackoverflow.com/questions/11117382/priority-in-go-select-statement-workaround

WhoSoup replied:
I was not aware of that, thank you!

PaulBernier commented on controller_routing.go:
```diff
@@ -4,9 +4,13 @@ import "time"
 // route takes messages from ToNetwork and routes it to the appropriate peers
 func (c *controller) route() {
+	c.logger.Debug("Start controller.route()")
+	defer c.logger.Debug("Start controller.route()")
```
incorrect log message

PaulBernier commented on network.go:
```diff
@@ -104,19 +106,33 @@ func (n *Network) SetMetricsHook(f func(pm map[string]PeerMetrics)) {
+func (n *Network) Stop() error {
+	select {
+	case <-n.stopper:
+		return fmt.Errorf("network not running")
```
"network already stopped" is slightly better wording? Because your wording would include the case where the network has not been started yet, but in this case it wouldn't go in this case (because stopper would still be open but empty).

PaulBernier commented on network.go:
```diff
@@ -27,8 +29,8 @@ type Network struct {
-	globalCloser chan interface{}
-	fatalError   chan error
+	stopper        chan interface{}
+	startStopMutex sync.Mutex
```
not using that?

WhoSoup replied:
I was playing around with the ability to stop and then restart the network but abandoned it for now since it would dramatically increase the complexity with little to no gain compared to just creating a new network.

PaulBernier commented on controller_connections.go:
```diff
@@ -295,17 +298,23 @@ func (c *controller) listen() {
+	defer tmpLogger.Debug("controller.listen() stopping")
 	c.listener = l
 
+	go func() { // the listener doesn't play well with immediately stopping
```
I saw you explained that in your PR comment but I still failed to understand why this can't be in the for loop, what problem does it have that this go routine approach doesn't have>

WhoSoup replied:
Edit: I remembered why

There are two issues at work here.

The first issue, why the check for c.net.stopper isn't in the for loop:

By default, the `listener.Accept` function will block forever. This is obviously unsuitable for a shutdown. There are two ways to interrupt the `listener.Accept` and those are A) closing the underlying listener, or B) giving it a deadline so it periodically times out. 

Since the Accept function isn't a channel, I can't enclose it in a `select {}` statement. I'd have to set a deadline, then wait for the timeout to happen, then check to see if the stopper channel is closed. That takes up time and as long as the listener is listening, the port on the machine is tied up and you can't start a new network on the same port. You could set the deadline to something like 1 second but that uses up unnecessary resources.

To get it to close as fast as possible, we want to close the listener from the outside.

----

The reason why I closed the listener in a goroutine is the very rare edge case of that if the network is stopped very quickly after being started, the variable `controller.listener` may still be nil and it's not possible to stop a goroutine that hasn't executed yet from running. 

With the current implementation, the goroutine will close the listener after it exists. We could add a second check at the beginning of the function that will prevent the listener from even being started if the network was stopped already but that might be overkill.

PaulBernier replied:
I understand better now, thank you.

WhoSoup commented:
Looks like the PR may have had a few Friday-Code symptoms. Sorry about that. I addressed the comments and reworked the awkward closing of the "manage online" loop.

The issue I wanted to avoid was that there are Peers who were in the middle of connecting when the network closed, that then keep around as "ghost" peers. They'd close eventually due to being unable to reply to "Pings" with a Pong and the remote node closing the connection, timing out, or channels filling up, all of which are less than ideal scenarios. 

The solution I went with is to check for `Network.stopper` in two places in the peer. Once in the sendloop, once in the Stop() function. The latter is there to avoid deadlocks where the peer tries to report its own status but the channel isn't being emptied. I'm not a big fan of checking the network stopper in the peer file but this is a solution that doesn't depend on processing all the events in the manage-online-channel after shutdown.

PaulBernier commented on peer.go:
```diff
@@ -86,8 +86,10 @@ func (p *Peer) Stop() {
 		// sendLoop closes p.conn and p.send in defer
-
-		p.net.controller.peerStatus <- peerStatus{peer: p, online: false}
+		select {
+		case p.net.controller.peerStatus <- peerStatus{peer: p, online: false}:
+		case <-p.net.stopper:
+		}
```
Shouldn't this rather be:
```golang
select {
  case <-p.net.stopper:
  default:
      p.net.controller.peerStatus <- peerStatus{peer: p, online: false}:
}
```
if `p.net.stopper` is not closed then proceed to the peer un-registration?

WhoSoup replied:
No, if it's written that way, the line inside default will block until a signal is received without reacting to the stopper channel

PaulBernier replied:
Sure, in case `peerStatus` would be full that could be a problem...

PaulBernier commented on peer.go:
```diff
@@ -199,6 +201,9 @@ func (p *Peer) sendLoop() {
 	defer p.conn.Close() // close connection on fatal error
 	for {
 		select {
+		case <-p.net.stopper:
+			p.Stop()
+			return
```
Shouldn't this logic be in its own go routine? Because it's not particularly related to Send logic, at least no more than read logic for instance. Also it's a bit cleaner and clearer if this SendLoop is exited through a single path (the `case <-p.stop`).

WhoSoup replied:
I think that's a fair request from a theoretical standpoint but from a practical one, having another long-term goroutine in the system that effectively adds nothing isn't necessary. I think the performance impact is negligible enough that if you feel it's worth it, I'll add it.

PaulBernier replied:
Nah it's fine.
